x-pack/filebeat/input/websocket/input.go

```diff
@@ -0,0 +1,438 @@
+// context cancellation or type invalidity errors, any other error will be retried.
+func (input) Run(env v2.Context, src inputcursor.Source, crsr inputcursor.Cursor, pub inputcursor.Publisher) error {
+	var cursor map[string]interface{}
+	if !crsr.IsNew() { // Allow the user to bootstrap the program if needed.
+		err := crsr.Unpack(&cursor)
+		if err != nil {
+			return err
+		}
+	}
+	return input{}.run(env, src.(*source), cursor, pub)
+}
+
+func (i input) run(env v2.Context, src *source, cursor map[string]interface{}, pub inputcursor.Publisher) error {
+	cfg := src.cfg
+	i.cfg = cfg
+	log := env.Logger.With("input_url", cfg.Resource.URL)
+
+	metrics := newInputMetrics(env.ID)
+	defer metrics.Close()
+	metrics.resource.Set(cfg.Resource.URL.String())
+	metrics.errorsTotal.Set(0)
+
+	ctx := ctxtool.FromCanceller(env.Cancelation)
+
+	patterns, err := regexpsFromConfig(cfg)
+	if err != nil {
+		metrics.errorsTotal.Inc()
+		return err
+	}
+
+	prg, ast, err := newProgram(ctx, cfg.Program, root, patterns, log)
+	if err != nil {
+		metrics.errorsTotal.Inc()
+		return err
+	}
+	var state map[string]interface{}
+	if cfg.State == nil {
+		state = make(map[string]interface{})
+	} else {
+		state = cfg.State
+	}
+	if cursor != nil {
+		state["cursor"] = cursor
+	}
+
+	// websocket client
+	headers := formHeader(cfg)
+	url := cfg.Resource.URL.String()
+	c, _, err := websocket.DefaultDialer.Dial(url, headers)
+	if err != nil {
+		metrics.errorsTotal.Inc()
+		log.Errorw("failed to establish websocket connection", "error", err)
+		return err
+	}
+	defer c.Close()
+
+	done := make(chan struct{})
+	errChan := make(chan error)
+
+	go func() {
+		defer close(done)
+		for {
+			_, message, err := c.ReadMessage()
+			if err != nil {
+				metrics.errorsTotal.Inc()
+				if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+					log.Debugw("websocket connection closed, attempting to reconnect", "error", err)
+					c, err = connectWebSocketWithRetry(log, url, cfg.Resource.Retry)
+					if err != nil {
+						log.Errorw("failed to reconnect websocket", "error", err)
+						errChan <- err
+						return
+					}
+					log.Debugw("reconnected to websocket")
+					continue
+				}
+				log.Errorw("failed to read websocket data", "error", err)
+				errChan <- err
+				return
+			}
+			metrics.receivedBytesTotal.Add(uint64(len(message)))
+			state["response"] = message
+			log.Debugw("received websocket message", logp.Namespace("websocket"), string(message))
+			err = i.processAndPublishData(ctx, metrics, prg, ast, state, cursor, pub, log)
+			if err != nil {
+				metrics.errorsTotal.Inc()
+				log.Errorw("failed to process and publish data", "error", err)
+				errChan <- err
+				return
+			}
+		}
+	}()
+
+	// blocks until done is closed , context is cancelled or an error is received
+	for {
+		select {
+		case err := <-errChan:
+			return err
+		case <-done:
+			return nil
```

Comment:
```suggestion
		case err := <-done:
			return err
```
(https://go.dev/play/p/0YgP84ykJyv)